### PR TITLE
Added support for alternate character encodings

### PR DESCRIPTION
**Pull request by hexane360**
This PR adds support for alternate UTF character encodings. It hews as closely as possible to the YAML spec, as described in [section 5.2](https://yaml.org/spec/1.2/spec.html#id2771184). In particular, this means that byte order marks (BOMs) are allowed at the beginning of any document, not just the beginning of a stream.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -6,6 +6,7 @@ version = "0.4.5"
 Base64 = "2a0f44e3-6c83-55bd-87e4-b1978d98bd5f"
 Dates = "ade2ca70-3891-5945-98fb-dc099432e06a"
 Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
+StringEncodings = "69024149-9ee7-55f6-a4c4-859efe599b68"
```
Comment (kescobo) [inline Project.toml:9]: Does this need a compat entry, or is it part of the StdLib?

File: src/scanner.jl [hunk 2]
```diff
@@ -1419,6 +1474,9 @@ function scan_plain_spaces(stream::TokenStream, indent::Integer,
                 forwardchars!(stream)
             else
                 push!(breaks, scan_line_break(stream))
+                if peek(stream.input) == '\uFEFF'
```
Comment (kescobo) [inline src/scanner.jl:1477]: Can you add something to test this branch?

**Comment by kescobo** [comment]
I don't know what this is for exactly, but if it gets us closer to YAML spec I'm all for it. I left a couple of naive comments, would love to get a more knowledgable reviewer if possible.

Also - any idea what's going on with v1.3 tests? I don't remember why we test on v1.3, but I have some vague memory that it's there for a reason, so I don't just want to nuke it.

File: Project.toml (hunk 1 above)
Comment (hexane360) [reply Project.toml:9]: It definitely needs a compat entry.

File: src/scanner.jl (hunk 2 above)
Comment (hexane360) [reply src/scanner.jl:1477]: Yeah, that's a good idea. IIRC it's triggered for a YAML document with just a bare string:
```
---
just a string
\uFEFF---
another document
```

**Comment by hexane360** [comment]
I'll need to look into what's going on with Julia 1.3. I suspect method resolution is behaving slightly differently.

I'd appreciate a more knowledgeable reviewer taking a look.

My goal with this is to be able to handle almost any valid YAML file in the wild. Maybe exotic UTF encodings are pretty rare, but there's still a lot of software (e.g. notepad.exe) that generates byte order marks, so handling those correctly is important.

**Review by kescobo — APPROVED** [review]

**Comment by hexane360** [comment]
Julia 1.3 was easier than expected; it turns out `io.jl` used the `peek` function differently.

Just now, I added a test case that should cover line 1477. It's needed for a document with a BOM after a document with trailing whitespace.

**Comment by kescobo** [comment]
Fantastic, thanks! I'm just going to wait for CI to complete and then merge :+1:

**Comment by hexane360** [comment]
I think the checks are stalled. Oddly enough though, it shows as completed in Github Actions: https://github.com/JuliaData/YAML.jl/actions/runs/504374788

**Comment by christopher-dG** [comment]
Restarted CI to see if that magically fixes anything, but I think the required check might be set up incorrectly (cc @kescobo).

**Comment by kescobo** [comment]
I don't understand why `julia 1` isn't completing... That's the only think I listed as being required. But that's identical right now to `julia 1.5`, right? I'll just force merge and remove the lock for now.

**Pull request merged**